js/common/lib/tensor-impl.ts

```diff
@@ -210,15 +266,153 @@ export class Tensor implements TensorInterface {
   }
   // #endregion
 
-  // #region fields
+  // #region public fields
   readonly dims: readonly number[];
   readonly type: TensorType;
-  readonly data: TensorDataType;
   readonly size: number;
   // #endregion
 
+  // #region private fields
+
+  /**
+   * stores the location of the data.
+   */
+  private dataLocation: TensorDataLocation;
+
+  /**
+   * stores the data on CPU, if location is 'cpu' or 'cpu-pinned'. otherwise empty.
+   */
+  private cpuData?: TensorDataType;
+
+  /**
+   * stores the underlying texture when location is 'texture'. otherwise empty.
+   */
+  private gpuTextureData?: TensorTextureType;
+
+  /**
+   * stores the underlying GPU buffer when location is 'gpu-buffer'. otherwise empty.
+   */
+  private gpuBufferData?: TensorGpuBufferType;
+
+  /**
+   * stores an optional downloader function to download data from GPU to CPU.
+   */
+  private downloader?(): Promise<TensorDataType>;
+
+  /**
+   * a flag indicating whether the data is being downloaded from GPU to CPU.
+   */
+  private isDownloading?: boolean;
+
+  /**
+   * stores an optional disposer function to dispose the underlying data.
+   */
+  private disposer?(): void;
+  // #endregion
+
+  // #region properties
+  get data(): TensorDataType {
+    this.ensureValid();
+    if (!this.cpuData) {
+      throw new Error(
+          'The data is not on CPU. Use `getData()` to download GPU data to CPU, ' +
+          'or use `texture` property to access the GPU data directly.');
```

Comment:
nit: or use `texture` or `gpuBuffer` (based on your EP) property to access the GPU data directly?